### PR TITLE
Revise interface of ComponentRegistry

### PR DESCRIPTION
**Pull request by IzumiSy**
Implementation for https://github.com/fabrix-framework/fabrix/issues/79

### Revised interface of `ComponentRegistry`.

* Add `getFabrixComponent` method: this is the method to create a custom component dynamically that composes the functionality internall that `FabrixComponent` is doing. The difference is that users can define `customProps` that is the extra props that the component callers can pass to the component (see how we can use the method in `packages/fabrix/__tests__/componentRegistry.test.tsx`)
* Introduced component categorization: **Unit** and **Composite**. Composite components are the component that are consisted of one or more unit components. Unit components are atomic components that are not composed of other components. `getFabrixComponent` only expects the name of **composite** components, because unit components don't expect data structure, but a field-level, atomic information.

### Another renderer: `FabrixCustomComponent`.

* This is the component that is similar to `FabrixComponent`, but the difference is that the component expects callers to pass `component` prop that is an infromation of a component to render with. `FabrixCustomComponent` itself is not exposed to end-users, instead `getFabrixComponent` method from `ComponentRegistry` automatically retrieves the entry information of a composite custom component from it with the given name from users. 
* I have also made changes in `renderer.tsx` in this PR to get some fundamental functions from `FabrixComponent` to share with `FabrixCustomComponent`, because those two should have the same interface as much as possible.

### Other
* Moved functions related to tables into `renders/table.tsx` as it was gathered in `renders/fields.tsx`

### CODE REVIEW

File: packages/fabrix/src/customRenderer.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+              component={{
+                name: props.component.name,
+                entry: componentEntry,
+                customProps: props.component.customProps,
+              }}
+            />
+          );
+        }
+        default: {
+          throw new Error(`Unsupported component type: ${componentEntry.type}`);
```
Comment (IzumiSy) [inline packages/fabrix/src/customRenderer.tsx:47]: Support for `views` and `form` will be another PR.

File: packages/fabrix/src/registry.tsx [hunk 2]
```diff
@@ -24,153 +31,260 @@ export type Field = {
...
+      isSubmitting: boolean;
+    }) => React.ReactElement,
+  ) => React.ReactNode;
+  renderReset: (
+    renderer: (props: { reset: () => void }) => React.ReactNode,
+  ) => React.ReactNode;
+};
+
 export type TableComponentHeader = Field & {
   render: ((rowValue: Record<string, unknown>) => React.ReactElement) | null;
```
Comment (toiroakr) [inline packages/fabrix/src/registry.tsx:83]: Is there a problem with ReactNode?
I have changed from ReactElement to ReactNode because I wanted to register a renderer that could return null before.
```suggestion
  render: ((rowValue: Record<string, unknown>) => React.ReactNode) | null;
```

File: packages/fabrix/src/registry.tsx [hunk 3]
```diff
@@ -24,153 +31,260 @@ export type Field = {
...
+  /**
+   * Get the component by name.
+   *
+   * If the component is not found, it will return the default component.
+   *
+   * @param name The name of the component.
+   * @param type The type of the component.
+   */
+  getCustomComponentByNameWithFallback<T extends ComponentEntries["type"]>(
+    name: string | null | undefined,
```
Comment (toiroakr) [inline packages/fabrix/src/registry.tsx:267]: Why not just a `string`?
```suggestion
    name: string
```

Comment (toiroakr) [reply packages/fabrix/src/registry.tsx:267]: Is it better?
```tsx
  getCustomComponentByNameWithFallback<
    T extends ComponentEntries["type"],
    N extends P["custom"] extends { composite: CompositeComponentMap }
      ? KeyOf<P["custom"]["composite"]>
      : string | null | undefined,
  >(name: N, type: T) {
```

File: packages/fabrix/src/registry.tsx [hunk 4]
```diff
@@ -24,153 +31,260 @@ export type Field = {
...
+
+  /**
+   * Get the component by name.
+   *
+   * If the component is not found, it will return the default component.
+   *
+   * @param name The name of the component.
+   * @param type The type of the component.
+   */
+  getCustomComponentByNameWithFallback<T extends ComponentEntries["type"]>(
```
Comment (toiroakr) [inline packages/fabrix/src/registry.tsx:266]: Even if it's just called `getCustomComponent`, it doesn't feel that strange.

File: packages/fabrix/src/registry.tsx (hunk 2 above)
Comment (IzumiSy) [reply packages/fabrix/src/registry.tsx:83]: No. This might be unexpectedly degraded by me 🙏

File: packages/fabrix/src/registry/utiltypes.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        : T extends "formField"
+          ? FormFieldComponentProps
+          : T extends "form"
+            ? FormComponentProps
+            : T extends "tableCell"
+              ? TableCellComponentProps
+              : T extends "table"
+                ? TableComponentProps
+                : never
+  >;
```
[the remark this answers is not in the document]
Comment (IzumiSy) [reply packages/fabrix/src/registry/utiltypes.ts:82]: Great!

File: packages/fabrix/src/registry.tsx (hunk 4 above)
Comment (IzumiSy) [reply packages/fabrix/src/registry.tsx:266]: OK, then let's go with `getCustomComponent`. 

I am also supposing users can tell the role of methods by method signatures 👍

Comment (IzumiSy) [reply packages/fabrix/src/registry.tsx]: https://github.com/fabrix-framework/fabrix/pull/93/commits/997e75c77f29efbadffb0fa9441e713a682831bd

File: packages/fabrix/src/registry.tsx (hunk 3 above)
Comment (IzumiSy) [reply packages/fabrix/src/registry.tsx]: https://github.com/fabrix-framework/fabrix/pull/93/commits/0638dbfe054ab831ecb99b43929a9c17c0b9e99b

**Review by toiroakr — APPROVED** [review]
